### ISSUE TITLE
Address parser not RFC 5321 compliant

### ISSUE THREAD

**Issue by gene-hightower** [haraka/Haraka#2833]
### system info

Linux (Fedora 31)

Haraka from git repo, head.

### Expected behaviour

Tried a few addresses from the Wikipedia page <https://en.wikipedia.org/wiki/Email_address#Examples>

### Observed behaviour

Tried some addresses from the "valid" list that failed, specifically quoted strings in the local part seem to be a problem area.

$ telnet localhost 2225
Trying ::1...
Connected to localhost.
Escape character is '^]'.
220 digilicious.com ESMTP Haraka/2.8.25 ready
EHLO digilicious.com
250-digilicious.com Hello localhost.localdomain [::1]Haraka is at your service.
250-PIPELINING
250-8BITMIME
250-SMTPUTF8
250 SIZE 0
MAIL FROM:<" "@DiGiLiCious.com>
501-Command parsing failed
501 Error: Invalid local part in address: " "@DiGiLiCious.com

Also, some of the invalid address seem to go just fine:

MAIL FROM:<just"not"right@DiGiLiCious.com>
250 sender <"just\"not\"right"@DiGiLiCious.com> OK

Should have failed, I think.

Also:

MAIL FROM:<this is"not\allowed@DiGiLiCious.com>
250 sender <"this\ is\"not\\allowed"@DiGiLiCious.com> OK

...and other, I think you get the idea.

Other areas that seemed to be a problem where UTF-8 in the domain part of an address. With SMTPUTF8 support that should work.

### Steps to reproduce

Just telnet to Haraka and speak SMTP.

I guess my question is, are you guys open to using a parser that follows the RFC-5321 mailbox syntax?

**Comment by msimerson** [comment]
> I guess my question is, are you guys open to using a parser that follows the RFC-5321 mailbox syntax?

Yes! We are. Currently we extracted the parser out of Haraka to [address-rfc2821](https://github.com/haraka/node-address-rfc2821). We've long had intentions of updating the parser for full 5321 compliance.

**Comment by gene-hightower** [comment]
I will take a shot at creating a parser for RFC-5321 syntax. Plan to try https://pegjs.org/ to generate from PEG.

**Comment by baudehlo** [comment]
I highly recommend taking a look at
https://github.com/jackbearheart/email-addresses/blob/master/lib/email-addresses.js
-
it uses a hand made parser, but it follows the grammar rules in structure.
